File: app/src/main/java/com/prey/activities/PermissionInformationActivity.java
/*******************************************************************************
 * Created by Carlos Yaconi
 * Copyright 2015 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities;

import android.Manifest;
import android.annotation.TargetApi;
import android.content.Intent;
import android.content.pm.ActivityInfo;

import com.prey.PreyConfig;
import com.prey.PreyPermission;

import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Window;

import androidx.core.app.ActivityCompat;

import com.prey.PreyLogger;
import com.prey.actions.aware.AwareController;
import com.prey.backwardcompatibility.FroyoSupport;
import com.prey.services.PreyOverlayService;

public class PermissionInformationActivity extends PreyActivity {

    private static final int SECURITY_PRIVILEGES = 10;
    private String congratsMessage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
    }

    @Override
    public void onBackPressed() {
    }

    @Override
    protected void onResume() {
        super.onResume();
        showScreen();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        PreyLogger.d("requestCode:" + requestCode + " resultCode:" + resultCode);
        if (requestCode == SECURITY_PRIVILEGES)
            showScreen();
    }

    private void showScreen() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PreyLogger.d("PermissionInformationActivity: Build.VERSION_CODES >=M");
            boolean canAccessFineLocation = PreyPermission.canAccessFineLocation(this);
            boolean canAccessCoarseLocation = PreyPermission.canAccessCoarseLocation(this);
            boolean canAccessCamera = PreyPermission.canAccessCamera(this);
            boolean canAccessStorage = PreyPermission.canAccessStorage(this);
            boolean configurated=canAccessFineLocation&&canAccessCoarseLocation && canAccessCamera
                    && canAccessStorage ;
            if(!configurated){
                askForPermission();
            }
        }
        if (FroyoSupport.getInstance(this).isAdminActive()) {
                Intent intentPermission = null;
                boolean canDrawOverlays = PreyPermission.canDrawOverlays(this);
                if(!canDrawOverlays) {
                    askForPermissionAndroid7();
                    startOverlayService();
                }else{
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        intentPermission = new Intent(PermissionInformationActivity.this, CheckPasswordHtmlActivity.class);
                    } else {
                        intentPermission = new Intent(PermissionInformationActivity.this, LoginActivity.class);
                    }
                }
                PreyConfig.getPreyConfig(PermissionInformationActivity.this).setProtectReady(true);
                new Thread() {
                    public void run() {
                        try{
                            AwareController.getInstance().init(getApplicationContext());
                        }catch(Exception e){
                            PreyLogger.e("Error:"+e.getMessage(),e);
                        }
                    }
                }.start();
                try {
                    if(intentPermission!=null) {
                        startActivity(intentPermission);
                    }
                }catch (Exception e){
                    PreyLogger.e("Error:"+e.getMessage(),e);
                }
                finish();
            } else {
                Intent intentPrivileges = FroyoSupport.getInstance(getApplicationContext()).getAskForAdminPrivilegesIntent();
                startActivityForResult(intentPrivileges, SECURITY_PRIVILEGES);
            }
    }
    @TargetApi(Build.VERSION_CODES.M)
    public void askForPermission() {
        ActivityCompat.requestPermissions(PermissionInformationActivity.this, INITIAL_PERMS, REQUEST_PERMISSIONS);
    }
    private static final int REQUEST_PERMISSIONS = 5;

    private static final String[] INITIAL_PERMS = {
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.CAMERA,
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
    };

    @TargetApi(Build.VERSION_CODES.M)
    public void askForPermissionAndroid7() {
        PreyLogger.d("PermissionInformationActivity: askForPermissionAndroid7");
        Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getPackageName()));
        startActivityForResult(intent, OVERLAY_PERMISSION_REQ_CODE);
        startOverlayService();
    }
    public static int OVERLAY_PERMISSION_REQ_CODE = 5473;

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        PreyLogger.d("PermissionInformationActivity: onRequestPermissionsResult");
        boolean canAccessFineLocation = PreyPermission.canAccessFineLocation(this);
        boolean canAccessCoarseLocation = PreyPermission.canAccessCoarseLocation(this);
        boolean canAccessCamera = PreyPermission.canAccessCamera(this);
        boolean canAccessStorage = PreyPermission.canAccessStorage(this);
        if (canAccessFineLocation && canAccessCoarseLocation && canAccessCamera
                && canAccessStorage  ) {
            boolean canDrawOverlays = PreyPermission.canDrawOverlays(this);
            if (!canDrawOverlays) {
                askForPermissionAndroid7();
                startOverlayService();
            } else {
                if (!canDrawOverlays) {
                    askForAdminActive();
                } else {
                        finish();
                        Intent intent = null;
                        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                            intent = new Intent(this, CheckPasswordHtmlActivity.class);
                        }else{
                            intent = new Intent(this, CheckPasswordActivity.class);
                        }
                        startActivity(intent);
                }
            }
        }
    }

    private void startOverlayService() {
        PreyLogger.d("PermissionInformationActivity: startOverlayService");
        Intent intentOverlay = new Intent(getApplicationContext(), PreyOverlayService.class);
        startService(intentOverlay);
    }

    public void askForAdminActive() {
        Intent intentPrivileges = FroyoSupport.getInstance(getApplicationContext()).getAskForAdminPrivilegesIntent();
        startActivityForResult(intentPrivileges, SECURITY_PRIVILEGES);
    }

}

File: app/src/main/java/com/prey/activities/PermissionInformationActivity.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.activities

import android.Manifest
import android.annotation.TargetApi
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.Window
import androidx.core.app.ActivityCompat

import com.prey.actions.aware.AwareController
import com.prey.backwardcompatibility.FroyoSupport
import com.prey.PreyConfig
import com.prey.PreyLogger
import com.prey.PreyPermission
import com.prey.services.PreyOverlayService
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Activity that displays information about the required permissions.
 *
 * This activity is responsible for explaining the necessary permissions to the user
 * and providing a way to grant them.
 */
class PermissionInformationActivity : PreyActivity() {
    private val congratsMessage: String? = null

    /**
     * Called when the activity is created.
     * Initializes the activity and sets up the window feature.
     *
     * @param savedInstanceState The saved instance state, or null if not saved
     */
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        requestWindowFeature(Window.FEATURE_NO_TITLE)
    }

    override fun onResume() {
        super.onResume()
        showScreen()
    }

    /**
     * Called when an activity you launched exits, giving you the requestCode you started it with,
     * the resultCode it returned, and any additional data from it.
     *
     * @param requestCode The request code of the activity that was launched
     * @param resultCode The result code of the activity that was launched
     * @param data The data returned by the activity that was launched
     */
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        PreyLogger.d("requestCode:$requestCode resultCode:$resultCode")
        if (requestCode == SECURITY_PRIVILEGES) showScreen()
    }

    /**
     * Shows the screen and performs necessary actions based on the device's configuration.
     */
    private fun showScreen() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PreyLogger.d("PermissionInformationActivity: Build.VERSION_CODES >=M")
            val canAccessFineLocation: Boolean = PreyPermission.canAccessFineLocation(this)
            val canAccessCoarseLocation: Boolean = PreyPermission.canAccessCoarseLocation(
                this
            )
            val canAccessCamera: Boolean = PreyPermission.canAccessCamera(this)
            val canAccessStorage: Boolean = PreyPermission.canAccessStorage(this)
            val configurated = (canAccessFineLocation && canAccessCoarseLocation && canAccessCamera
                    && canAccessStorage)
            if (!configurated) {
                askForPermission()
            }
        }
        if (FroyoSupport.getInstance(this).isAdminActive()) {
            var intentPermission: Intent? = null
            val canDrawOverlays: Boolean = PreyPermission.canDrawOverlays(this)
            if (!canDrawOverlays) {
                askForPermissionAndroid7()
                startOverlayService()
            } else {
                intentPermission = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    Intent(
                        this@PermissionInformationActivity,
                        CheckPasswordHtmlActivity::class.java
                    )
                } else {
                    Intent(this@PermissionInformationActivity, LoginActivity::class.java)
                }
            }
            PreyConfig.getInstance(this@PermissionInformationActivity).setProtectReady(true)
            CoroutineScope(Dispatchers.IO).launch {
                try {
                    AwareController.getInstance().initLastLocation(applicationContext)
                } catch (e: Exception) {
                    PreyLogger.e("Error: ${e.message}", e)
                }
            }
            try {
                if (intentPermission != null) {
                    startActivity(intentPermission)
                }
            } catch (e: Exception) {
                PreyLogger.e("Error: ${e.message}", e)
            }
            finish()
        } else {
            val intentPrivileges =
                FroyoSupport.getInstance(applicationContext).getAskForAdminPrivilegesIntent()
            startActivityForResult(intentPrivileges, SECURITY_PRIVILEGES)
        }
    }

    /**
     * Asks for the necessary permissions.
     */
    @TargetApi(Build.VERSION_CODES.M)
    fun askForPermission() {
        PreyConfig.getInstance(applicationContext).setActivityView(PERMISSIONS_ASK)
        ActivityCompat.requestPermissions(
            this@PermissionInformationActivity,
            INITIAL_PERMS,
            REQUEST_PERMISSIONS
        )
    }

    /**
     * Asks for the Android 7 overlay permission.
     * This permission is required for the app to draw over other apps.
     */
    @TargetApi(Build.VERSION_CODES.M)
    fun askForPermissionAndroid7() {
        PreyLogger.d("PermissionInformationActivity: askForPermissionAndroid7")
        val intent = Intent(
            Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse(
                "package:$packageName"
            )
        )
        startActivityForResult(intent, OVERLAY_PERMISSION_REQ_CODE)
        startOverlayService()
    }

    /**
     * Handles the result of the permission request.
     * This method is called after the user has granted or denied the requested permissions.
     *
     * @param requestCode The request code of the permission request
     * @param permissions The array of requested permissions
     * @param grantResults The array of grant results for the requested permissions
     */
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        PreyLogger.d("PermissionInformationActivity: onRequestPermissionsResult")
        val canAccessFineLocation: Boolean = PreyPermission.canAccessFineLocation(this)
        val canAccessCoarseLocation: Boolean = PreyPermission.canAccessCoarseLocation(this)
        val canAccessCamera: Boolean = PreyPermission.canAccessCamera(this)
        val canAccessStorage: Boolean = PreyPermission.canAccessStorage(this)
        if (canAccessFineLocation && canAccessCoarseLocation && canAccessCamera
            && canAccessStorage
        ) {
            val canDrawOverlays: Boolean = PreyPermission.canDrawOverlays(this)
            if (!canDrawOverlays) {
                askForPermissionAndroid7()
                startOverlayService()
            } else {
                if (!canDrawOverlays) {
                    askForAdminActive()
                } else {
                    finish()
                    var intent: Intent? = null
                    intent = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        Intent(this, CheckPasswordHtmlActivity::class.java)
                    } else {
                        Intent(this, CheckPasswordActivity::class.java)
                    }
                    startActivity(intent)
                }
            }
        }
    }

    /**
     * Starts the overlay service.
     * This service is responsible for drawing over other apps.
     */
    private fun startOverlayService() {
        PreyLogger.d("PermissionInformationActivity: startOverlayService")
        val intentOverlay = Intent(applicationContext, PreyOverlayService::class.java)
        startService(intentOverlay)
    }

    /**
     * Asks for admin privileges.
     * This method is called when the app needs to perform administrative tasks.
     */
    fun askForAdminActive() {
        val intentPrivileges =
            FroyoSupport.getInstance(applicationContext).getAskForAdminPrivilegesIntent()
        startActivityForResult(intentPrivileges, SECURITY_PRIVILEGES)
    }

    companion object {
        const val SECURITY_PRIVILEGES = 10
        const val REQUEST_PERMISSIONS = 5
        const val OVERLAY_PERMISSION_REQ_CODE: Int = 5473
        const val PERMISSIONS_ASK: String = "PERMISSIONS_ASK"
        private val INITIAL_PERMS = arrayOf(
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION,
            Manifest.permission.CAMERA,
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
        )
    }

}